src/estimators/advi.jl

```diff
@@ -0,0 +1,57 @@
+
+struct ADVI{Tlogπ, B <: Union{Function, Bijectors.Inverse{<:Bijectors.Bijector}}} <: AbstractGradientEstimator
+    # Automatic differentiation variational inference
+    # 
+    # Kucukelbir, A., Tran, D., Ranganath, R., Gelman, A., & Blei, D. M. (2017).
+    # Automatic differentiation variational inference.
+    # Journal of machine learning research.
+
+    ℓπ::Tlogπ
```

Comment:
This is IMO annoying because it ties the target distribution/model to the method (ADVI, in this case). 

IMO, a `GradientEstimator` should not be tied to the VI-approach _by default_. It should indeed be possible to overload the estimation of a particular object, e.g. `ELBO`, for any given VI approach, e.g. `ADVI`, but for example the `ELBO` is much more generally applicable than _just_ `ADVI` (in fact, `ADVI` is just special because it makes use of a closed-form expression for the entropy term of the base-distribution).

Reply:
In general, it's also much more preferable to have something like
```julia
estimate_gradient!(vi_method, objective, target, ...)
```
than taking a wrapper-struct approach, e.g.
```julia
estimate_gradient(VIMethod(objective, target), ...)
```

Reply:
> IMO, a `GradientEstimator` should not be tied to the VI-approach _by default_. It should indeed be possible to overload the estimation of a particular object, e.g. `ELBO`, for any given VI approach, e.g. `ADVI`, but for example the `ELBO` is much more generally applicable than _just_ `ADVI` (in fact, `ADVI` is just special because it makes use of a closed-form expression for the entropy term of the base-distribution).

I'm not sure about this, because, if one changes the objective, the algorithm for estimation has to entirely change in general. For instance, something like alpha divergence minimization will share little code with ADVI as far as I can see. So why not just reflect this inherent coupling in the interface?

Reply:
> I'm not sure about this, because, if one changes the objective, the algorithm for estimation has to entirely change in general. For instance, something like alpha divergence minimization will share little code with ADVI as far as I can see. So why not just reflect this inherent coupling in the interface?

Sure, but what about the scenario of me wanting to minimize the ELBO but using an empirical estimate to the entropy-term rather than the closed-form expression, which is used in ADVI? Clearly both of these are estimating the ELBO, but now I need to re-implement the ELBO just for this new method rather than taking advantage of the existing implementation for `ADVI`, just overriding the way the entropy is computed. Does that make sense?

Reply:
> > I'm not sure about this, because, if one changes the objective, the algorithm for estimation has to entirely change in general. For instance, something like alpha divergence minimization will share little code with ADVI as far as I can see. So why not just reflect this inherent coupling in the interface?
> 
> Sure, but what about the scenario of me wanting to minimize the ELBO but using an empirical estimate to the entropy-term rather than the closed-form expression, which is used in ADVI? Clearly both of these are estimating the ELBO, but now I need to re-implement the ELBO just for this new method rather than taking advantage of the existing implementation for `ADVI`, just overriding the way the entropy is computed. Does that make sense?

I originally intended the `ADVI` terminology to strictly mean the closed-form entropy case, as that is what is done in the original paper. I did think about a name like `ClosedFromEntropyELBO` (although this is a mouthful) would be more appropriate. 

As per your argument, your example (Monte-Carlo estimated entropy) would be a different implementation of the ELBO estimator no? Then, wouldn't that require a separate implementation of the ELBO anyways? I'm certainly open to the idea of splitting the objective and the estimation method, but to me, it seems like there are a few cases where different objectives can share the estimation method, apart from Monte-Carlo entropy v.s. closed-form entropy.

Reply:
> I did think about a name like ClosedFromEntropyELBO (although this is a mouthful) would be more appropriate.

But why not just make it `elbo` and then inside `elbo` we call `estimate_entropy`, etc.? Then if you want a method that computes the entropy in closed form, e.g. ADVI, you just need to implement `estimate_entropy` rather than the entire thing. This seems strictly better, no?

> I'm certainly open to the idea of splitting the objective and the estimation method, but to me, it seems like there are a few cases where different objectives can share the estimation method, apart from the case of ELBO with Monte-Carlo entropy v.s. closed-form entropy.

But maybe you want to do CAVI, which _also_ uses the ELBO. Aaaand maybe you want to add variance reduction to all of these. 

Are we going to encapsulate all these approaches in the first iteration of whatever we end up with? Probably not :upside_down_face: 

_But_ because we don't quite know the full scope, I think it's always useful to define a very general interface allowing us to share as much code as possible, and then we can eventually just specialize further, e.g. the case of `Cholesky` as you've mentioned, when it makes sense.

Reply:
> As per your argument, your example (Monte-Carlo estimated entropy) would be a different implementation of the ELBO estimator no? Then, wouldn't that require a separate implementation of the ELBO anyways?

I didn't explicit answer this, but no it wouldn't if the ELBO is further separated into components which are re-usable. But I'm not saying we have to do all this separation from the very beginning, but I do prefer leaving it as an option for later if we see there is room for code-sharing. Hence why I'm not a big fan of hiding the objective as a closure in the gradient-computation itself.